packages/loopring_v3/contracts/aux/liquidity-providers/BasicLiquidityProvider.sol

```diff
@@ -0,0 +1,160 @@
+        address from;
+        address to;
+        address token;
+        uint96  amount;
+        uint32  storageID;
+        uint64  validUntil; // most significant 32 bits as block height, least significant 32 bits as block time
+        bytes   signature;  // provider's signature
+    }
+
+    bytes32 constant public FASTWITHDRAWAL_APPROVAL_TYPEHASH = keccak256(
+        "FastWithdrawalApproval(address exchange,address from,address to,address token,uint96 amount,uint32 storageID,uint64 validUntil)"
+    );
+
+    bytes32 public DOMAIN_SEPARATOR;
+
+    FastWithdrawalAgent public agent;
+
+    constructor(FastWithdrawalAgent _agent)
+    {
+        agent = _agent;
+        DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("BasicLiquidityProvider", "1.0", address(this)));
+    }
+
+    // User initiated fast withdrawal
+    function execute(FastWithdrawalApproval calldata approval)
+        external
+        nonReentrant
+    {
+        // Compute the hash
+        bytes32 hash = EIP712.hashPacked(
+            DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encodePacked(
+                    FASTWITHDRAWAL_APPROVAL_TYPEHASH,
+                    approval.exchange,
+                    approval.from,
+                    approval.to,
+                    approval.token,
+                    approval.amount,
+                    approval.storageID,
+                    approval.validUntil
+                )
+            )
+        );
+        // Check the signature
+        require(hash.verifySignature(owner, approval.signature), "INVALID_SIGNATURE");
+        require(checkValidUntil(approval.validUntil), "FASTWITHDRAWAL_APPROVAL_EXPIRED");
+
+        FastWithdrawalAgent.Withdrawal[] memory withdrawals = new FastWithdrawalAgent.Withdrawal[](1);
+        withdrawals[0] = FastWithdrawalAgent.Withdrawal({
+            exchange: approval.exchange,
+            from: approval.from,
+            to: approval.to,
+            token: approval.token,
+            amount: approval.amount,
+            storageID: approval.storageID
+        });
+        executeInternal(withdrawals);
+    }
+
+    // LP initiated fast withdrawals
+    function execute(FastWithdrawalAgent.Withdrawal[] memory withdrawals)
+        public
+        nonReentrant
+        onlyOwner
+    {
+        executeInternal(withdrawals);
+    }
+
+    // Allows the LP to transfer funds back out of this contract.
+    function transferOut(
+        address to,
+        address token,
+        uint    amount
+        )
+        external
+        nonReentrant
+        onlyOwner
+    {
+        if (amount > 0) {
+            if (token == address(0)) {
+                to.sendETHAndVerify(amount, gasleft()); // ETH
+            } else {
+                token.safeTransferAndVerify(to, amount);  // ERC20 token
+            }
+        }
+    }
+
+    // Allows the LP to enable the necessary ERC20 approvals
+    function approve(
+        address token,
+        address spender,
```

Comment:
is the spender usually the exchange contract?

Reply:
Only the agent if just used for the fast withdrawals.